File: interaccion/src/empaquetador_nodo.cpp
/**
 * Se implementa un nodo que espera recibir mensajes cuyo topic es "inf_pers_topic" del tipo interaccion::inf_personal_usuario.
 * Muestra en pantalla este mensaje recibido
 */

//Se incluyen las cabeceras
#include "ros/ros.h"
#include "interaccion/inf_personal_usuario.h"
#include "interaccion/usuario.h"
#include "std_msgs/String.h"
#include <string>

//Se declara el namespace std
using namespace std;

//Estructura para la informacion personal
struct {
    string nombre;
    int edad;
    std::vector<std::string> idiomas;
} infPersonal;


//Estructura para la posicion
struct {
    int x;
    int y;
    int z;
} posicion;

//Algunos bools de confirmacion de mensaje
bool confirmacionInfPersonal = false;
bool confirmacionEmocion = false;
bool confirmacionPosicion = false;

//Se declara el published como variable global
ros::Publisher publicadorEmpaquetador;

//Se instancia el mensaje a publicar como variable global
interaccion::usuario mensajeEmpaquetador;

/**
* Esta función muestra por pantalla el mensaje recibido que es de tipo inf_personal_usuario
*/
void funcionCallback(const interaccion::inf_personal_usuario::ConstPtr& msg){
 ROS_INFO("El nombre es: %s", msg->nombre.c_str());
 ROS_INFO("La edad es: %d", msg->edad);
 for (int i = 0; i < msg->idiomas.size(); i++) {
   ROS_INFO("Los idiomas que se son: %s", msg->idiomas[i].c_str());
 }
 
 //Se confirma la informacion personal
 confirmacionInfPersonal = true;
 ROS_INFO("confirmo info personal: %d", confirmacionInfPersonal);

 /*Se introduce la informacion personal recibida en la estructura infPersonal,
 y en el mensajeEmpaquetador*/
 mensajeEmpaquetador.infPersonal.edad = msg->edad;
 mensajeEmpaquetador.infPersonal.nombre = msg->nombre;
 mensajeEmpaquetador.infPersonal.idiomas = msg->idiomas;
}

/**
* Esta función muestra por pantalla el mensaje recibido que es de tipo String
*/
void funcionCallback2(const std_msgs::String::ConstPtr& msg){
 ROS_INFO("La emocion es: %s", msg->data.c_str());
 
 //Se confirma que llega el mensaje de la emocion
 confirmacionEmocion = true;
 ROS_INFO("confirmo info emocion: %d", confirmacionEmocion);
 
 //Se introduce el mensaje en el mensajeEmpaquetador
 mensajeEmpaquetador.emocion = msg->data.c_str();
}

/**
* Esta función muestra por pantalla el mensaje recibido que es de tipo pos_usuario
*/
void funcionCallback3(const interaccion::pos_usuario::ConstPtr& msg){
 ROS_INFO("La posicion en x es: %d", msg->x);
 ROS_INFO("La posicion en y es: %d", msg->y);
 ROS_INFO("La posicion en z es: %d", msg->z);
 
 //Se confirma que ha llegado el mensaje de la posicion
 confirmacionPosicion = true;
 ROS_INFO("confirmo info posicion: %d", confirmacionPosicion);
 
 /*Se introduce el mensaje recibido en la estructura posicion,
 y de ahi al mensajeEmpaquetador*/
 mensajeEmpaquetador.posicion.x = msg->x;
 mensajeEmpaquetador.posicion.y = msg->y;
 mensajeEmpaquetador.posicion.z = msg->z;
}

//Funcion principal
int main(int argc, char **argv){

 //registra el nombre del nodo: empaquetador_nodo y el handle empaquetadorNodo
 ros::init(argc, argv, "empaquetador_nodo");

 ros::NodeHandle empaquetadorNodo;

 //Creamos un objeto nodo
 ROS_INFO("empaquetador_nodo creado y registrado"); //muestra en pantalla

 //es necesario "advertir" el tipo de mensaje a enviar con el topic user_topic
 publicadorEmpaquetador = empaquetadorNodo.advertise<interaccion::usuario>("user_topic",0);

 //si recibimos el mensaje cuyo topic es: "inf_pers_topic" llamamos a la función manejadora: funcionCallback
 ros::Subscriber subscriptor = empaquetadorNodo.subscribe("inf_pers_topic", 0, funcionCallback);

 //si recibimos el mensaje cuyo topic es: "emocion_topic" llamamos a la función manejadora functionCallback2
 ros::Subscriber subscriptor2 = empaquetadorNodo.subscribe("emocion_topic", 0, funcionCallback2);

 //si recibimos el mensaje cuyo topic es: "pos_usuario_topic" llamamos a la función manejadora functionCallback3
 ros::Subscriber subscriptor3 = empaquetadorNodo.subscribe("pos_usuario_topic", 0, funcionCallback3);

 while (ros::ok()){
    //Cuando haya llegado toda la informacion, se publica el mensaje
    if (confirmacionEmocion == true and confirmacionInfPersonal == true and confirmacionPosicion == true)
     {
        publicadorEmpaquetador.publish(mensajeEmpaquetador);

        //Se reseteran para empaquetar nueva informacion
        confirmacionEmocion = false;
        confirmacionInfPersonal = false;
        confirmacionPosicion = false;
    }
    ros::spinOnce();
 }
}

File: interaccion/src/empaquetador_nodo.py
#CARGAMOS ROSPY Y LOS MENSAJES
import rospy
from trabajo.msg import emocion_usuario, inf_personal_usuario,pos_usuario,usuario
from std_msgs.msg import Bool,String

#CALLBACK QUE LLAMA CUANDO LLEGA LA INFORMACION PERSONAL
def callback_uno(data):
    #IMPRIMIMOS QUE HA LLEGADO LA INFORMACION, PARA TENER UN CONTROL DEL PASO DE MENSAJES MAS DIRECTO
    rospy.loginfo("Llega la informacion personal")
    #DEFINIMOS VARIABLES GLOBALES QUE NECESITAMOS
    global flag_enviado,flag_info,usuarioT,lista_nombre,lista_edad,lista_idiomas

    #AÑADIMOS A LAS LISTAS LA INFORMACION QUE HA LLEGADO
    lista_nombre.append(str(data.nombre))
    lista_edad.append(int(data.edad))
    lista_idiomas.append(list(data.idiomas))

    #ESTABLECEMOS LA INFORMACION A ENVIAR COMO LA INFORMACION EN LA PRIMERA POSICION DE LA LISTA
    usuarioT.inf.nombre=lista_nombre[0]
    usuarioT.inf.edad=lista_edad[0]
    usuarioT.inf.idiomas=lista_idiomas[0]

    #INDICAMOS QUE HEMOS CARGADO LA INFORMACION
    flag_info=True

#CALLBACK QUE LLAMA CUANDO LLEGA LA POSICION
def callback_dos(data):
    #IMPRIMIMOS QUE HA LLEGADO LA POSICION, PARA TENER UN CONTROL DEL PASO DE MENSAJES MAS DIRECTO
    rospy.loginfo("llega posicion")

    #DEFINIMOS VARIABLES GLOBALES QUE NECESITAMOS
    global flag_enviado,flag_pos,usuarioT,lista_x,lista_y,lista_z

    #AÑADIMOS A LAS LISTAS LA INFORMACION QUE HA LLEGADO
    lista_x.append(int(data.x))
    lista_y.append(int(data.y))
    lista_z.append(int(data.z))

    #ESTABLECEMOS LA INFORMACION A ENVIAR COMO LA INFORMACION EN LA PRIMERA POSICION DE LA LISTA
    usuarioT.posicion.x=lista_x[0]
    usuarioT.posicion.y=lista_y[0]
    usuarioT.posicion.z=lista_z[0]

    #INDICAMOS QUE HEMOS CARGADO LA INFORMACION
    flag_pos=True

#CALLBACK QUE LLAMA CUANDO LLEGA LA EMOCION DE LA PERSONA
def callback_tres(data):
    #IMPRIMIMOS QUE HA LLEGADO LA EMOCION, PARA TENER UN CONTROL DEL PASO DE MENSAJES MAS DIRECTO
    rospy.loginfo("llega emocion")

    #DEFINIMOS VARIABLES GLOBALES QUE NECESITAMOS
    global flag_enviado,flag_emo,usuarioT,lista_emo

    #AÑADIMOS A LAS LISTAS LA INFORMACION QUE HA LLEGADO
    lista_emo.append(str(data))

    #ESTABLECEMOS LA INFORMACION A ENVIAR COMO LA INFORMACION EN LA PRIMERA POSICION DE LA LISTA
    usuarioT.emocion=lista_emo[0]

    #INDICAMOS QUE HEMOS CARGADO LA INFORMACION
    flag_emo=True



def listener():
    #INICIALIZAMOS EL NODO
    rospy.init_node('empaquetador', anonymous=True)
    #DEFINIMOS LOS SUBSCRIPTORES
    rospy.Subscriber('inf_pers_topic', inf_personal_usuario,callback_uno)
    rospy.Subscriber('pos_usuario_topic', pos_usuario,callback_dos)
    rospy.Subscriber('emocion_topic', String, callback_tres)
    #DEFINIMOS LAS VARIABLES GLOBALES QUE VAMOS USANDO
    global lista_nombre,flag_enviado,lista_edad,lista_idiomas,lista_x,lista_y,lista_z,lista_emo,flag_emo,flag_pos,flag_info



if __name__ == '__main__':
    #DEFINIMOS COMO GLOBALES LAS VARIABLES QUE VAMOS USANDO
    global lista_nombre,lista_edad,lista_idiomas,lista_x,lista_y,lista_z,lista_emo
    #INICIALIZAMOS LAS VARIABLES QUE VAMOS A UAR
    lista_nombre=[]
    lista_edad=[]
    lista_idiomas=[]
    lista_x=[]
    lista_y=[]
    lista_z=[]
    lista_emo=[]
    flag_emo=False
    flag_pos=False
    flag_info=False
    flag_enviado=False
    #INICIAMOS EL MENSAJE QUE SE VA A ENVIAR, DEL TIPO usuario
    usuarioT=usuario()  
    #INICIAMOS LA ESCUCHA
    listener()
    #INICIAMOS EL PUBLICADOR
    pub = rospy.Publisher('user_topic', usuario,queue_size=1)

    while not rospy.is_shutdown():
        #SI SE RECIBE LA INFORMACION DE LOS 3 NODOS
        if flag_emo and flag_info and flag_pos:

            try:
                #ENVIA LA INFORMACION Y RESETEA LAS FLAGS
                rospy.loginfo("enviando")
                flag_enviado=True
                if len(lista_emo)<=1:
                   flag_emo=False
                if len(lista_z)<=1:
                    flag_pos=False
                if len(lista_nombre)<=1:
                    flag_info=False

                pub.publish(usuarioT)

                #BORRAMOS LA INFORMACION DE LA PRIMERA POSICION DE LAS LISTAS, QUE ES LA QUE YA HEMOS ENVIADO
                lista_x.pop(0)
                lista_y.pop(0)
                lista_z.pop(0)
                lista_emo.pop(0)
                lista_nombre.pop(0)
                lista_edad.pop(0)
                lista_idiomas.pop(0)
                if len(lista_x)>0:
                    usuarioT.x=lista_x[0]
                if len(lista_y)>0:
                    usuarioT.y=lista_y[0]
                if len(lista_z)>0:
                    usuarioT.z=lista_z[0]
                if len(lista_nombre)>0:
                    usuarioT.nombre=lista_nombre[0]
                if len(lista_edad)>0:
                    usuarioT.edad=lista_edad[0]
                if len(lista_emo)>0:
                    usuarioT.emocion=lista_emo[0]
                if len(lista_idiomas)>0:
                    usuarioT.idiomas=lista_idiomas[0]

            except rospy.ROSInterruptException:
                pass
